---
title: "Project-1(BRC)"
author: "Aastha Guragain"
date: "2025-01-24"
output:
  html_document:
    df_print: paged
  pdf_document: default
---



Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
```{r}
#Setting the working directory
setwd("~/Desktop/seurat-learning")

#Loading the necessary Libraries
#install.packages("tinytex")
#tinytex::install_tinytex()
library(harmony)
library(ggplot2)
library(Seurat)
library(SeuratObject)
library(patchwork)
```

```{r}
# Reading file
brc_file <- readRDS("breast_carcinoma.rds")
#Extracting metadata
metadata.brc <- brc_file@meta.data
```

```{r}
#Calculating Mitochondrial percentage and adding it to metadata if not given
brc_file$mpercent <- PercentageFeatureSet(brc_file, pattern = "^MT")
metadata.brc$mpercent <- brc_file$mpercent
View(metadata.brc)

```

```{r}
# In case nCount_RNA and n_Feature RNA not given:
#Extracting count matrix
count_matrix <- brc_file@assays$RNA@counts

#creating seurat object from count mmatrix
seurat_obj_cnt <- CreateSeuratObject(count_matrix)
new_seu_metadata <- seurat_obj_cnt@meta.data

#setting the variable
n_feature <- new_seu_metadata$nFeature_RNA
n_count <- new_seu_metadata$nCount_RNA
```


```{r}
#Incorporating ncount and nfeature information into raw suerat file expecting it do not have these features
brc_file$n_feature <- n_feature
brc_file$n_count <- n_count
brc_file$new_mito_percent <- PercentageFeatureSet(brc_file, pattern = "^MT-")

#Incorporating into metadata
metadata.brc$n_feature <- n_feature
metadata.brc$n_count <- n_count
metadata.brc$new_mito_percent <- PercentageFeatureSet(brc_file, pattern = "^MT-")

```

```{r}
#Upstream Analysis#------------------------------------------------------------------

# Visualize QC metrics as a violin plot
VlnPlot(brc_file, features = c("nCount_RNA", "nFeature_RNA", "percent_mito"), ncol = 3)
```
```{r}
#A scatter plot visualizing the relationship between the number of detected features (nFeature_RNA) and the total RNA count (nCount_RNA). 
plot1 <- FeatureScatter(brc_file, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot2 <- FeatureScatter(brc_file, feature1 = "nCount_RNA", feature2 = "percent_mito")
plot1 + plot2
```



```{r}
#Creating a filtered subset of the brc_file dataset (subset_brc), retaining cells with 700–4000 detected features (nFeature_RNA), 500–20,000 total RNA counts (nCount_RNA), and less than 10% mitochondrial content (percent_mito)
subset_brc <- subset(brc_file, subset = nFeature_RNA <4000 & nFeature_RNA > 700 & nCount_RNA < 20000 & nCount_RNA >500 & percent_mito <10)
View(subset_brc@meta.data)
```


```{r}
#Violin plot displaying the distribution of filtered nCount RNA, nFeature RNA and mitochondrial percentage across cells
plot1 = VlnPlot(subset_brc, features = c("nCount_RNA", "nFeature_RNA", "percent_mito"), ncol = 3)
plot1
```


```{r}
#Feature scatter plot
plot2 <- FeatureScatter(subset_brc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot3 <- FeatureScatter(subset_brc, feature1 = "nCount_RNA", feature2 = "percent_mito")
plot2 + plot3
```
```{r}
#Normalization to isolate the true biological variation from technical artifacts and accurately compare gene expression patterns
subset_brc <- NormalizeData(subset_brc, normalization.method = "LogNormalize", scale.factor = 10000)

```

```{r}
#Finding variable features
subset_brc <- FindVariableFeatures(subset_brc, selection.method = "vst", nfeatures = 2000)
```


```{r}
#Calculating the top10 variable genes
top10 <- head(VariableFeatures(subset_brc), 10)
top10
```

```{r}
# Plot variable features without label and with label
plot1 <- VariableFeaturePlot(subset_brc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)

# Display the plot
plot1 + plot2

```

```{r}
#Scaling the data to remove unwanted sources of variation from a single-cell dataset
all.genes <- rownames(subset_brc)
View(all.genes)
subset_brc <- ScaleData(subset_brc, features = all.genes)

```


```{r}
#Perform linear dimensional reduction (PCA)
subset_brc <- RunPCA(subset_brc, features = VariableFeatures(object = subset_brc))

#Examine and visualize PCA results a few different ways
VizDimLoadings(subset_brc, dims = 1:2, reduction = "pca")
```


```{r}
#Dimensional reduction plot, each point represents the cell
DimPlot(subset_brc, reduction = "pca") + NoLegend()
```

```{r}
#Heatmap for easy exploration of the primary sources of heterogeneity
DimHeatmap(subset_brc, dim = 1, cells =500, balanced = TRUE)
```


```{r}
#Determining the top PCs that represents the robust compression of the dataset
ElbowPlot(subset_brc)
```


```{r}
#Clustering the cells
#First constructing a KNN graph based on the euclidean distance in PCA space, and refining the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). 
subset_brc <- FindNeighbors(subset_brc, dims = 1:15) #dims = dimension, 
subset_brc <- FindClusters(subset_brc, resolution = 0.5)
DimPlot(subset_brc, reduction = "pca")
```


```{r}
#Uniform Manifold Approximation and Projection (UMAP), performed on the filtered dataset (subset_brc) using the first 15 principal components (dims = 1:15) from PCA. 
subset_brc <- RunUMAP(subset_brc, dims = 1:15, reduction = "pca")

#Cell clustering based on seurat_clusters
DimPlot(subset_brc, reduction = "umap", group.by = "seurat_clusters")
```


```{r}
#Cell clustering based on disease
disease <- DimPlot(subset_brc, reduction = "umap", group.by = "disease" , label = "TRUE")
disease
```

```{r}
#Cell clustering based on batch run
batch <- DimPlot(subset_brc, reduction = "umap", group.by = "batch_run", label = "TRUE")
batch
```


```{r}
#Cell clustering based on cell type
cell_type<- DimPlot(subset_brc, reduction = "umap", group.by = "cell_type")
cell_type
```


```{r}
#Batch effect removal using harmony
subset_brc <- RunHarmony(
  object = subset_brc,
  group.by.vars = "batch_run",
  dims.use = 1:15 # Use the same dimensions you selected in FindNeighbors
)

```

```{r}
# Re- running clustering using harmony
subset_brc <- subset_brc %>%
  RunUMAP(reduction = 'harmony',dims = 1:15) %>% 
  FindNeighbors(reduction = 'harmony' , dims = 1:15) %>%
  FindClusters(resolution = 0.5)
```





```{r}
disease_after_harmony <- DimPlot(subset_brc, reduction = "umap", group.by = "disease" , label = "TRUE")
batch_after_harmony <- DimPlot(subset_brc, reduction = "umap", group.by = "batch_run", label = "TRUE")
disease
```


```{r}
disease_after_harmony
```


```{r}
batch
```


```{r}
batch_after_harmony
```








